### DISCUSSION TITLE
What does actually happen with the return value of `LightningModule.validation_step`?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [Lightning-AI/pytorch-lightning#19305] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
`LightningModule.validation_step` operates on a single batch of validation data and typically returns a validation loss computed from that single batch. According to the [docs](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#validation), when validation happens, we iterate through the complete validation data set and compute the loss for every batch. What happens to the computed losses?

Intuitively, it would make sense to average them (weighted by the actual batch size which can be smaller for the last batch) so that they are comparable to the training loss computed in the last training step. But the documentation is not explicit about this. I can follow the source code only to to [here](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/loops/evaluation_loop.py#L396), where I think we get the output of `validation_step` (if the stars and program state align). Then I get lost, when the output is passed to callback hook functions that again dynamically determine what function they are going to call with what arguments...

Edit: Wait, do I have to look in the `log` method instead?
